lib/HLSL/DxilContainerReflection.cpp

```diff
@@ -180,16 +182,14 @@ class DxilShaderReflection : public DxilModuleReflection, public ID3D12ShaderRef
   DXC_MICROCOM_TM_ADDREF_RELEASE_IMPL()
   DXC_MICROCOM_TM_CTOR(DxilShaderReflection)
   HRESULT STDMETHODCALLTYPE QueryInterface(REFIID iid, void **ppvObject) override {
-    HRESULT hr = DoBasicQueryInterface<ID3D12ShaderReflection>(this, iid, ppvObject);
-    if (hr == E_NOINTERFACE) {
-      // ID3D11ShaderReflection is identical to ID3D12ShaderReflection, except
-      // for some shorter data structures in some out parameters.
-      PublicAPI api = IIDToAPI(iid);
-      if (api == m_PublicAPI) {
-        *ppvObject = (ID3D12ShaderReflection *)this;
-        this->AddRef();
-        hr = S_OK;
-      }
+    HRESULT hr = E_NOINTERFACE;
+    PublicAPI api = IIDToAPI(iid);
+    // ID3D11ShaderReflection is identical to ID3D12ShaderReflection, except
+    // for some shorter data structures in some out parameters.
+    if (api == m_PublicAPI || IsEqualIID(__uuidof(IUnknown), iid)) {
```

Comment:
This is my ignorance of what QI is meant to do. Given IUnknown, it should just return whatever it has because everything matches. Is that right?

Reply:
You should be able to QI `IUnknown` on any interface derived from such.  I'm not sure whether that answers your question, but then maybe I don't know what you're getting at.  In order to use the `IUnknown` result beyond the interface methods there, you'll have to QI for another interface, at which point we check that the interface matches our state.

Reply:
If it is IUnknown then we should be doing `*ppvObject = (IUnknown*)this`.

Reply:
I've found the "ID3D11ShaderReflection is identical to ID3D12ShaderReflection, except..." comment extremely confusing while reading this.  This implementation of QueryInterface is pretty complicated and unusual compared to most others you'd encounter implementing COM objects.

I think some key points here:
* although everything uses the same vtable, as from `ID3D12ShaderReflection`, there are differences in behavior depending on the API version, and there are 3 of these - it's not just d3d11 vs d3d12.
* when the object is created the API version is fixed
* from that point on, this object can only be QI'd for the matching API version.

Reply:
> If it is IUnknown then we should be doing *ppvObject = (IUnknown*)this.

What difference could this make without weird multiple inheritance situations?

> I've found the "ID3D11ShaderReflection is identical to ID3D12ShaderReflection, except..." comment extremely confusing while reading this.

Would you be satisfied with the point-form explanation you just wrote instead?

Reply:
IMO, too much of this code is based on (currently valid) assumptions about equivalent vtable layouts of various different interfaces that have appeared in various different iterations of the SDK.  I think there's value in idiomatic, non-surprising code.

I think the explanation you wrote in another thread helped capture it as well, but some acknowledge in the code that this is not your normal implementation of a COM object would be helpful.